src/Fabulous/StackAllocatedCollections.fs

```diff
@@ -1,132 +1,581 @@
+        | Few (struct (size, v0, v1, v2)) ->
+            match size with
+            | Size.Zero -> one(v)
+            | Size.One -> two(v0, v)
+            | Size.Two -> three(v0, v1, v)
+            | Size.Three -> many([| v0; v1; v2; v |])
+            | _ -> empty() // should never happen but don't want to throw there
+        | Many arr -> many(Array.appendOne v arr)
+
+
+    let inline length (arr: StackArray3<'v> inref) : int =
+        match arr with
+        | Few (struct (size, _, _, _)) -> int size
+        | Many arr -> arr.Length
+
+
+    let get (arr: StackArray3<'v> inref) (index: int) : 'v =
+        match arr with
+        | Few (struct (size, v0, v1, v2)) ->
+            if (index >= int size) then
+                IndexOutOfRangeException() |> raise
+            else
+                match index with
+                | 0 -> v0
+                | 1 -> v1
+                | _ -> v2
+
+        | Many arr -> arr.[index]
+
+
+    let find (test: 'v -> bool) (arr: StackArray3<'v> inref) : 'v =
+        match arr with
+        | Few (struct (size, v0, v1, v2)) ->
+            match (size, test v0, test v1, test v2) with
+            | Size.One, true, _, _
+            | Size.Two, true, _, _
+            | Size.Three, true, _, _ -> v0
+            | Size.Two, false, true, _
+            | Size.Three, false, true, _ -> v1
+            | Size.Three, false, false, true -> v2
+            | _ -> KeyNotFoundException() |> raise
+        | Many arr -> Array.find test arr
+
+
+    /// Note that you should always use the result,
+    /// In Few mode it creates a new stack allocated array
+    /// In Many case it sorts the Many variant inline for optimization reasons
+    let rec inline sortInPlace<'T, 'V when 'V: comparison>
+        ([<InlineIfLambda>] getKey: 'T -> 'V)
+        (arr: StackArray3<'T> inref)
+        : StackArray3<'T> =
+        match arr with
+        | Few (struct (size, v0, v1, v2)) ->
+            match size with
+            | Size.Zero
+            | Size.One -> arr
+            | Size.Two ->
+                if (getKey v0 > getKey v1) then
+                    two(v1, v0)
+                else
+                    arr
+            | Size.Three ->
+                match (getKey v0, getKey v1, getKey v1) with
+                // abc acb bac bca cba cab
+
+                //  a, c, b
+                | a, b, c when a <= c && c <= b -> three(v0, v2, v1)
+
+                //  b, a, c
+                | a, b, c when b <= a && a <= c -> three(v1, v0, v2)
+
+                //  b, c, a
+                | a, b, c when b <= c && c <= a -> three(v1, v2, v0)
+
+                //  c, b, a
+                | a, b, c when c <= b && b <= a -> three(v2, v1, v0)
+
+                //  c, a, b
+                | a, b, c when c <= a && a <= b -> three(v2, v0, v1)
+
+                // a, b, c left, thus already sorted
+                | _ -> arr
+
+
+            | _ -> empty() // should never happen but don't want to throw there
```

Comment:
Why not throw? I think it would be better to raise an exception users can see and report, rather than an odd behavior no one can explain

Reply:
I might be wrong about this, but even the function doesn't throw you pay perf cost at runtime that it **could** throw.

However, I don't remember where I got this idea, and not sure if it is actually true